libpromises/lastseen.c

```diff
@@ -766,3 +766,44 @@ int RemoveKeysFromLastSeen(const char *input, bool must_be_coherent,
 
     return 0;
 }
+
+/**
+ * @brief Acknowledge that lastseen host entry is observed.
+ * @param host_key The host key of the lastseen entry.
+ * @param already_acked Whether the lastseen entry was already acknowledged.
+ * @return true if host entry was successfully acknowledged, otherwise false.
+ */
+bool LastSeenHostAcknowledge(const char *host_key, bool incoming, bool *already_acked)
+{
+    DBHandle *db = NULL;
+    if (!OpenDB(&db, dbid_lastseen))
+    {
+        Log(LOG_LEVEL_ERR, "Unable to open lastseen DB");
+        return false;
+    }
+
+    // Update quality-of-connection entry
+    char key[CF_BUFSIZE];
+    NDEBUG_UNUSED int ret = snprintf(key, CF_BUFSIZE, "q%c%s", incoming ? 'i' : 'o', host_key);
+    assert(ret > 0 && ret < CF_BUFSIZE);
+
+    KeyHostSeen value;
+    if (!ReadDB(db, key, &value, sizeof(value)))
+    {
+        Log(LOG_LEVEL_ERR, "Unable to read key '%s' from lastseen DB", key);
+        CloseDB(db);
+        return false;
+    }
+
+    *already_acked = value.acknowledged;
+    value.acknowledged = true;
+
+    if (!WriteDB(db, key, &value, sizeof(value))) {
+        Log(LOG_LEVEL_ERR, "Unable to write key '%s' to lastseen DB", key);
+        CloseDB(db);
+        return false;
+    }
```

Comment:
I think this should happen in a transaction, i.e. it should use [`OverwriteDB()`](https://github.com/cfengine/core/blob/master/libpromises/dbm_api.h#L94)

Reply:
I don't understand this function. Please let me know if I'm using it wrong

Reply:
You could use the `OverwriteCondition` callback to only rewrite the value if it is being changed and to store the old value. That way you wouldn't have to do `ReadDB()` at all and there would be no writes in case there's no change.